abi-versions/v0.2.1/README.md

```diff
@@ -0,0 +1,1807 @@
+Indicates to the host that the plugin is done processing active
+context.
+
+This should be used after returning `false` in [`proxy_on_done`].
+
+Returned `status` value is:
+- `OK` on success.
+- `NOT_FOUND` when active context was not pending finalization.
+
+
+### `proxy_set_effective_context`
+
+* params:
+  - `i32 (uint32_t) context_id`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Changes the effective context to `context_id`.
+
+This can be used to change active context, e.g. during
+[`proxy_on_http_call_response`], [`proxy_on_grpc_receive`]
+and/or [`proxy_on_queue_ready`] callbacks.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `context_id`.
+
+
+## Configuration
+
+### `proxy_on_vm_start`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (size_t) vm_configuration_size`
+* returns:
+  - `i32 (bool) status`
+
+Called when the host starts the WebAssembly Virtual Machine.
+
+Its configuration (of `vm_configuration_size`) can be retrieved using
+[`proxy_get_buffer_bytes`] with `buffer_id` set to `VM_CONFIGURATION`.
+
+Plugin must return one of the following values:
+- `true` to indicate that the configuration was processed successfully.
+- `false` to indicate that the configuration processing failed.
+
+
+### `proxy_on_configure`
+
+* params:
+  - `i32 (uint32_t) context_id`
+  - `i32 (size_t) plugin_configuration_size`
+* returns:
+  - `i32 (bool) status`
+
+Called when the host starts the Proxy-Wasm plugin.
+
+Its configuration (of `plugin_configuration_size`) can be retrieved
+using [`proxy_get_buffer_bytes`] with `buffer_id` set to
+`PLUGIN_CONFIGURATION`.
+
+Plugin must return one of the following values:
+- `true` to indicate that the configuration was processed successfully.
+- `false` to indicate that the configuration processing failed.
+
+
+## Logging
+
+### `proxy_log`
+
+* params:
+  - `i32 (`[`proxy_log_level_t`]`) log_level`
+  - `i32 (const char *) message_data`
+  - `i32 (size_t) message_size`
+* returns:
+  - `i32 (`[`proxy_status_t`]`) status`
+
+Logs `message` at the `log_level`.
+
+Returned `status` value is:
+- `OK` on success.
+- `BAD_ARGUMENT` for unknown `log_level`.
+- `INVALID_MEMORY_ACCESS` when `message_data` and/or `message_size`
+  point to invalid memory address.
+
+
+### `wasi_snapshot_preview1.fd_write`
+
+* params:
+  - `i32 (`[`wasi_fd_id_t`]`) fd_id`
+  - `i32 (wasi_iovec_array_t) iovec`
+  - `i32 (size_t) iovec_size`
+  - `i32 (size_t *) return_written_bytes`
+* returns:
+  - `i32 (`[`wasi_errno_t`]`) errno`
+
+Logs message passed in `iovec`.
+
+When `fd_id` is `STDOUT`, then the message is going to be logged
+at the `INFO` level.
+
+When `fd_id` is `STDERR`, then the message is going to be logged
+at the `ERROR` level.
+
+Returned `errno` value is:
+- `SUCCESS` on success.
+- `BADF` for unknown or unsupported `fd_id`.
+- `FAULT` when `iovec`, `iovec_size` and/or `return_written_bytes`
+  point to invalid memory address.
+
+
+### `proxy_get_log_level`
+
```

Comment:
comment this is a guest function, better to but with the host in the same section and just clarify each?